cmd/feature.go

```diff
@@ -0,0 +1,194 @@
+
+It can also take a list of flags to update.
+
+If you give a null value, the flag will be removed.
+`,
+	Example: `$ cozy-stack feature flags --domain cozy.tools:8080 '{"add_this_flag": true, "remove_this_flag": null}'`,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		if flagDomain == "" {
+			errPrintfln("%s", errMissingDomain)
+			return cmd.Usage()
+		}
+		c := newAdminClient()
+		req := request.Options{
+			Method: "GET",
+			Path:   fmt.Sprintf("/instances/%s/feature/flags", flagDomain),
+		}
+		if len(args) > 0 {
+			req.Method = "PATCH"
+			req.Body = strings.NewReader(args[0])
+		}
+		res, err := c.Req(&req)
+		if err != nil {
+			return err
+		}
+		defer res.Body.Close()
+		var obj map[string]interface{}
+		if err = json.NewDecoder(res.Body).Decode(&obj); err != nil {
+			return err
+		}
+		for k, v := range obj {
+			fmt.Printf("- %s: %v\n", k, v)
+		}
+		return nil
+	},
+}
+
+var featureSetCmd = &cobra.Command{
+	Use:   "sets",
+	Short: `Display and update the feature sets for an instance`,
+	Long: `
+cozy-stack feature sets displays the feature sets coming from the manager.
+
+It can also take a list of sets that will replace the previous list (no merge).
+
+All the sets can be removed by setting an empty list ('').
+`,
+	Example: `$ cozy-stack feature sets --domain cozy.tools:8080 'set1 set2'`,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		if flagDomain == "" {
+			errPrintfln("%s", errMissingDomain)
+			return cmd.Usage()
+		}
+		c := newAdminClient()
+		req := request.Options{
+			Method: "GET",
+			Path:   fmt.Sprintf("/instances/%s/feature/sets", flagDomain),
+		}
+		if len(args) > 0 {
+			list := args
+			if len(args) == 1 {
+				list = strings.Fields(args[0])
+			}
+			buf, err := json.Marshal(list)
+			if err != nil {
+				return err
+			}
+			req.Method = "PUT"
+			req.Body = bytes.NewReader(buf)
+		}
+		res, err := c.Req(&req)
+		if err != nil {
+			return err
+		}
+		defer res.Body.Close()
+		var sets []string
+		if err = json.NewDecoder(res.Body).Decode(&sets); err != nil {
+			return err
+		}
+		for _, set := range sets {
+			fmt.Printf("- %v\n", set)
+		}
+		return nil
+	},
+}
+
+var featureContextCmd = &cobra.Command{
+	Use:   "context <context-name>",
+	Short: `Display and update the feature flags for a context`,
+	Long: `
+cozy-stack feature defaults displays the feature flags for a context.
+
+It can also create, update, or remove flags (with a ratio and value).
+
+To remove a flag, set it to an empty array (or null).
+`,
+	Example: `$ cozy-stack feature context beta '{"set_this_flag": [{"ratio": 0.1, "value": 1}, {"ratio": 0.9, "value": 2}] }'`,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		if len(args) == 0 || args[0] == "" {
+			return cmd.Usage()
+		}
+		c := newAdminClient()
+		req := request.Options{
+			Method: "GET",
+			Path:   fmt.Sprintf("/instances/feature/contexts/%s", args[0]),
+		}
+		if len(args) > 1 {
```

Comment:
IMHO you should check that the sum of ratio is 1 to at least return a warning

Reply:
Well, we can have a sum of ratio < 1: if no values is selected for this flag, it will use the default from `cozy-stack features defaults`

Reply:
This is why I talked about a warning and not an error. I don't think it's a common use case to define ratio sum < 1, at least the admin user doing this should be warned about it I think

Reply:
> I don't think it's a common use case to define ratio sum < 1

When discussed with Éric, it was the common case: the ratio is used on the feature flag activated (with a default for not being activated).

Reply:
ok